juju/bundle.py

```diff
@@ -156,311 +171,663 @@ def resolve(self, reference):
+                setattr(self, v, d[k])
+            else:
+                setattr(self, v, None)
+
+
+class AddApplicationChange(ChangeInfo):
+    _toPy = {'charm': 'charm',
+             'series': 'series',
+             'application': 'application',
+             'options': 'options',
+             'constraints': 'constraints',
+             'storage': 'storage',
+             'devices': 'devices',
+             'endpoint-bindings': 'endpoint_bindings',
+             'resources': 'resources',
+             'num-units': 'num_units'}
+
+    """
+    AddApplicationChange holds a change for deploying a Juju application.
+
+    :charm: holds the URL of the charm to be used to deploy this application.
+    :series: holds the series of the application to be deployed if the charm
+        default is not sufficient.
+    :application: holds the application name.
+    :num_units: holds the number of units required. For IAAS models, this will
+        be 0 and separate AddUnitChanges will be used. For Kubernetes models,
+        this will be used to scale the application.
+    :options: holds application options.
+    :constraints: holds the optional application constraints.
+    :storage: holds the optional storage constraints.
+    :devices: holds the optional devices constraints.
+    :endpoint_bindings: holds the optional endpoint bindings
+    :resources: identifies the revision to use for each resource of the
+        application's charm.
+    :local_resources: identifies the path to the local resource of the
+        application's charm.
+    """
+    def __init__(self, change_id, requires, params=None):
+        super(AddApplicationChange, self).__init__(change_id, requires)
+
+        if isinstance(params, list):
+            self.charm = params[0]
+            self.series = params[1]
+            self.application = params[2]
+            self.options = params[3]
+            self.constraints = params[4]
+            self.storage = params[5]
+            self.endpoint_bindings = params[6]
+            if len(params) == 8:
+                # Juju 2.4 and below only sends the resources
+                self.resources = params[7]
+                self.devices = None
+                self.num_units = None
+            else:
+                # Juju 2.5+ sends devices before resources, as well as num_units
+                # There might be placement but we need to ignore that.
+                self.devices = params[7]
+                self.resources = params[8]
+                self.num_units = params[9]
+
+        elif isinstance(params, dict):
+            AddApplicationChange.from_dict(self, params)
+        else:
+            raise Exception("unexpected params type")
+
+    @staticmethod
+    def method():
+        """
+        method returns an associated ID for the Juju API call.
+        """
+        return "deploy"
+
+    async def run(self, context):
+        """
+        Run executes a AddApplicationChange using the returned parameters from
```

Comment:
I'd drop "Run" and just leave it as "executes..." since the function name sets the context.